ion/integration_test.go

```diff
@@ -0,0 +1,808 @@
+}
+
+var nonEquivsSkipList = []string{
+	"bools.ion",
+	"decimals.ion",
+	"documents.ion",
+	"floats.ion",
+	"floatsVsDecimals.ion",
+	"localSymbolTableWithAnnotations.ion",
+	"structs.ion",
+	"symbolTables.ion",
+	"symbolTablesUnknownText.ion",
+	"symbols.ion",
+	"timestamps.ion",
+}
+
+func TestBinaryRoundTrip(t *testing.T) {
+	readFilesAndTest(t, goodPath, binaryRoundTripSkipList, func(t *testing.T, path string) {
+		testBinaryRoundTrip(t, path)
+	})
+}
+
+func TestTextRoundTrip(t *testing.T) {
+	readFilesAndTest(t, goodPath, textRoundTripSkipList, func(t *testing.T, path string) {
+		testTextRoundTrip(t, path)
+	})
+}
+
+func TestLoadBad(t *testing.T) {
+	readFilesAndTest(t, badPath, malformedIonsSkipList, func(t *testing.T, path string) {
+		testLoadBad(t, path)
+	})
+}
+
+func TestEquivalency(t *testing.T) {
+	readFilesAndTest(t, equivsPath, equivsSkipList, func(t *testing.T, path string) {
+		testEquivalency(t, path, true)
+	})
+}
+
+func TestNonEquivalency(t *testing.T) {
+	readFilesAndTest(t, nonEquivsPath, nonEquivsSkipList, func(t *testing.T, path string) {
+		testEquivalency(t, path, false)
+	})
+}
+
+// Execute round trip testing for BinaryWriter. Create a BinaryWriter, using the BinaryWriter create
+// a TextWriter and back to BinaryWriter again. Validate that the first and last Writers are equal.
+func testBinaryRoundTrip(t *testing.T, fp string) {
+	fileBytes := loadFile(t, fp)
+
+	// Make a binary writer from the file
+	binWriter, buf := encodeAsBinaryIon(t, fileBytes)
+
+	// Re-encode binWriter's stream as text into a string builder
+	_, str := encodeAsTextIon(t, buf.String())
+
+	// Re-encode str as binary in another binary writer
+	binWriter2, _ := encodeAsBinaryIon(t, []byte(str.String()))
+
+	// Compare the 2 binary writers
+	if !reflect.DeepEqual(binWriter, binWriter2) {
+		t.Errorf("Round trip test failed on: " + fp)
+	}
+}
+
+// Execute round trip testing for TextWriter. Create a TextWriter, using the TextWriter creat a
+// BinaryWriter and back to TextWriter again. Validate that the first and last Writers are equal.
+func testTextRoundTrip(t *testing.T, fp string) {
+	fileBytes := loadFile(t, fp)
+
+	// Make a text writer from the file
+	txtWriter, str := encodeAsTextIon(t, string(fileBytes))
+
+	// Re-encode txtWriter's stream as binary into a bytes.Buffer
+	_, buf := encodeAsBinaryIon(t, []byte(str.String()))
+
+	// Re-encode buf's stream as text in another text writer
+	txtWriter2, _ := encodeAsTextIon(t, buf.String())
+
+	//compare the 2 text writers
+	if !reflect.DeepEqual(txtWriter, txtWriter2) {
+		t.Errorf("Round trip test failed on: " + fp)
+	}
+}
+
+// Create a TextWriter from data parameter. Return the writer and string builder containing writer's contents
+func encodeAsTextIon(t *testing.T, data string) (Writer, strings.Builder) {
+	reader := NewReader(strings.NewReader(data))
+	str := strings.Builder{}
+	txtWriter := NewTextWriter(&str)
+	writeToWriterFromReader(t, reader, txtWriter)
+	txtWriter.Finish()
```

Comment:
Only caught this through linting, I'm getting a few warnings with unhandled exceptions.

Been looking through best practices for error handling in tests and it seems we can just make this obvious if we're ignoring an error by doing `_ = txtWriter.Finish()`.

Looking into what the best practices are, but I'd feel failing early when an error occurred would be good. Let says if the text writer failed to finish, would this be obvious if we continued to ignoring the error and let the test fail later on? But, this could be a unit test vs integration test scenario, so maybe explicity ignoring the error for these cases is good enough.

Thoughts?

Reply:
Correct, I did not put error handling for ```writer.Finish()```, ```stepIn()``` and ```stepOut()```.
My approach was to test the testfiles against different scenarios of this file (roundtrip, equivalency, loadBad). And if it fails at any point, the test fails. But open to suggestions.

Reply:
My general preference would be to explicitly check for errors from the writer and `t.Fatal(err)`; if something goes wrong during writing, that'll give you more information about the problem than one of the outputs being truncated.

The writers _should_ remember any errors they encounter along the way, no-op future writes, and return the error from `Finish`, so in theory you can get away with just checking the return value there. (Although that's maybe something that's worth testing more explicitly than it is tested now. :))

Reply:
Done.